HeterogeneousCore/CUDACore/src/CUDAESManaged.cc

```diff
@@ -0,0 +1,30 @@
+#include "HeterogeneousCore/CUDACore/interface/CUDAESManaged.h"
+
+
+CUDAESManaged::CUDAESManaged(): prefetched_(false) {}
+
+CUDAESManaged::~CUDAESManaged() {
+  for(auto& ptrSize: buffers_) {
+    cudaFree(ptrSize.first);
+  }
+}
+
+void CUDAESManaged::advise() const {
+  for(const auto& ptrSize: buffers_) {
+    cudaCheck(cudaMemAdvise(ptrSize.first, ptrSize.second, cudaMemAdviseSetReadMostly, 0)); // device is ignored for this advise
+  }
+}
+
+void CUDAESManaged::prefetchAsync(cuda::stream_t<>& stream) const {
+  // The boolean atomic is an optimization attempt, it doesn't really
+  // matter if more than one thread/edm stream issues the prefetches
+  // as long as most of the prefetches are avoided.
+  if(prefetched_.load())
+    return;
```

Comment:
Looking through the code again I noticed that here we/I have a single flag to control behaviour for all devices. I think this should be changed to either
1. the `atomic<bool>` flag should be changed to per-device flags, or
2. issue the prefetch for all devices below

In principle the latter (2) could make more sense as we expect all devices to run all code (and thus need all GPU conditions) anyway. But async prefetch needs also a CUDA stream, so it needs a bit more thought. Basically it would mean that the CUDAESManaged would have additional CUDA streams (one per device), and then use CUDA events to "synchronize" with the argument stream (otherwise there would be no gain from async). Then the `atomic<bool>` flag would actually be replaced with `cudaEventQuery()` for the CUDA event (of a device). I would naively imagine `atomic<bool>` to be faster than `cudaEventQuery()`, which would favor option 1.